aten/src/ATen/native_parse.py

```diff
@@ -18,72 +18,106 @@
 # type translations to continue to emit the legacy func schema for further
 # processing by downstream tools. This will helps us avoid having to prematurely
 # change all downstream tools to detect these new types.
-def temp_type_translations(typ):
-    # Enables Tensor[] by translating to legacy TensorList. See [temp translations]
-    if typ == 'Tensor[]':
-        return 'TensorList'
-    # Enables int[] by translating to legacy IntList. See [temp translations]
-    if typ == 'int[]':
-        return 'IntList'
-    # Enables int by translating to legacy int64_t. See [temp translations]
-    if typ == 'int':
-        return 'int64_t'
-    # Enables float by translating to legacy double. See [temp translations]
-    if typ == 'float':
-        return 'double'
-    return typ
-
-
-def parse_default(s):
-    if s.lower() == 'true':
-        return True
-    elif s.lower() == 'false':
-        return False
-    elif s == 'nullptr':
-        return s
+def type_argument_translations(arg):
```

Comment:
This function is basically trying to do everything.

IMO the way you want to structure this is so you can just rip out the legacy translations, but there's all sorts of things in here, like legacy sanitation (why?), default handling, size parsing, etc.  How about something like this:

1) something that just does try argument legacy translations.  This only needs to look at the type specifier.  It doesn't need to pull out the numbers inside `[]` -- you have to do that in whether legacy translation exists or not, it shouldn't be in here.
2) something that does legacy default translation.  It needs to have the type because the translations are type specific.  But you can have it be based on the pre-translated (i.e. correct) type, so if you remove the specific type translation, nothing has to change here.

Then leave everything else the same, but write the other parts to operate on pre-translated types, so instead of parsing the `3` out of `IntList[3]` you parse it out of `int[3]`.  The translations are just for passing to downstream components, native_parse should be "clean" otherwise.

Reply:
In defense of this PR as is, it does deduplicate two code paths. That seems like a win in my book. So I'd be OK with trying to refactor this code even more as we keep working on this path, while committing this for now.

Reply:
The distinction between legacy translations and non-legacy translations is not clear. Once the func schema complies with the JIT signature schema, we will start ripping out pretty much all the translations. That will happen one by one. Whether that corresponds to a legacy translation or something before then won't matter much. It is however important to make it very easy to understand what and how exactly things are being mapped. Separating this into legacy and non-legacy will make that more difficult to understand.

Reply:
regarding @ezyang's point about deduplicating two code paths, that's true although the code paths should at least be partially separate (return arguments can't have defaults, now they are parsed the same way and we just pretend the default specification wasn't there.